### PR TITLE
fix(gw): directory URL normalization

### PR DESCRIPTION
**Pull request by lidel**
## tl;dr

This PR fixes a regression around directory listing and index.html hosting,
and makes Kubo behavior compliant with desired [Gateway Specs here](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#use-in-directory-url-normalization).

### Why

Seems that during one of recent refactors code changed and we no longer check for trailing slash in HTTP request path.  Most likely, this happened when we switched from raw URL path to content path.

This regression meant Kubo was not following specs at
 [/ipfs/specs/http-gateways/PATH_GATEWAY.md#use-in-directory-url-normalization](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#use-in-directory-url-normalization).

### What

This PR: 

- cleans things up so we always redirect directories from `/ipfs/cid/foo` to `/ipfs/cid/foo/`
  - ensures dir behavior is the same  for  both index.html hosting and dir-index-html (generated listing).
  - maintains behavior described in https://github.com/ipfs/kubo/pull/3963 + adds test for it (we had none :see_no_evil:)
  - changes redirect code from 302 to 301 (Permanent)
- adds more tests, so we catch any future regressions.

### CODE REVIEW

File: test/sharness/t0115-gateway-dir-listing.sh [hunk 1]
```diff
@@ -43,8 +43,14 @@ test_expect_success "path gw: backlink on root CID should be hidden" '
   test_should_not_contain "<a href=\"/ipfs/$DIR_CID/\">..</a>" list_response
 '
 
-test_expect_success "path gw: Etag should be present" '
+test_expect_success "path gw: redirect dir listing to URL with traling slash" '
```
Comment (iand) [inline test/sharness/t0115-gateway-dir-listing.sh:46]:
```suggestion
test_expect_success "path gw: redirect dir listing to URL with trailing slash" '
```

File: test/sharness/t0115-gateway-dir-listing.sh [hunk 2]
```diff
@@ -72,31 +78,37 @@ test_expect_success "path gw: hash column should be a CID link with filename par
...
 
-test_expect_success "path gw: backlink on root CID should be hidden" '
+test_expect_success "subdomain gw: backlink on root CID should be hidden" '
   curl -sD - --resolve $DIR_HOSTNAME:$GWAY_PORT:127.0.0.1 http://$DIR_HOSTNAME:$GWAY_PORT/ > list_response &&
   test_should_contain "Index of" list_response &&
   test_should_not_contain "<a href=\"/\">..</a>" list_response
 '
 
-test_expect_success "path gw: Etag should be present" '
+test_expect_success "subdomain gw: redirect dir listing to URL with traling slash" '
```
Comment (iand) [inline test/sharness/t0115-gateway-dir-listing.sh:87]:
```suggestion
test_expect_success "subdomain gw: redirect dir listing to URL with trailing slash" '
```

File: test/sharness/t0115-gateway-dir-listing.sh [hunk 3]
```diff
@@ -121,8 +133,14 @@ test_expect_success "dnslink gw: backlink on root CID should be hidden" '
   test_should_not_contain "<a href=\"/\">..</a>" list_response
 '
 
-test_expect_success "dnslink gw: Etag should be present" '
+test_expect_success "dnslink gw: redirect dir listing to URL with traling slash" '
```
Comment (iand) [inline test/sharness/t0115-gateway-dir-listing.sh:136]:
```suggestion
test_expect_success "dnslink gw: redirect dir listing to URL with trailing slash" '
```

**Review by iand — APPROVED** [review]
LGTM, some minor typos in tests
